### PR TITLE
[cli] Add --active-arch-only flag

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
Closes ENG-11220
[Feature Request](https://expo.canny.io/feature-requests/p/support-reactnativearchitectures-in-expo-cli)
RFC for adding the `--active-arch-only` flag. This flag is used to reduce build times on android when using the new architecture by only building the abis of connected devices. I've kept the name the same so there is nothing new if you are coming from the community CLI. 

# How
Query the connected devices for their CPU architectures and pass this to gradle to override the `reactNativeArchitectures` gradle property.

### Edit 1:
- Change the flag to `--all-arch` and remove from help text
- Make the behaviour of building only for connected devices the default
- 
### Edit 2:
- If we have a device, only look up its abi instead of all connected devices
- Only check the architectures if we are building a debug variant

# Test Plan
Tested on a new blank expo project. 
- Enable new architecture
- Run the build with and without the flag

Without the flag: APK Size: 157.4MB, Build Time: 40.27s
With the flag: APK Size: 47.9MB, Build Time: 10.87s

Both tests were run as a clean build.

### CODE REVIEW

**Comment by brentvatne** [comment]
> Without the flag: APK Size: 157.4MB, Build Time: 40.27s
> With the flag: APK Size: 47.9MB, Build Time: 10.87s

nice ;)

**Review by brentvatne — APPROVED** [review]

File: packages/@expo/cli/src/run/android/index.ts [hunk 1]
```diff
@@ -47,6 +48,7 @@ export const expoRunAndroid: Command = async (argv) => {
     --variant <name>       Build variant. {dim Default: debug}
     -d, --device [device]  Device name to run the app on
     -p, --port <port>      Port to start the dev server on. {dim Default: 8081}
+    --active-arch-only     Only build for the architectures of connected devices or emulators
```
Comment (EvanBacon) [inline packages/@expo/cli/src/run/android/index.ts:51]: I think we should probably turn this feature on by default and possibly add an environment variable to disable setting `PreactNativeArchitectures` or configure the exact value.

Comment (janicduplessis) [reply packages/@expo/cli/src/run/android/index.ts:51]: I also agree it should be on by default. And yes this feature is meant to be used only for dev builds.

**Comment by Kudo** [comment]
thanks @alanjhughes for helping this task. i would reassign my linear task for you. close ENG-11220.
some thought from me:
- would be good to add the feature parity on ios. on ios, xcode has default `ONLY_ACTIVE_ARCH=YES` on debug build but not release build. if we could add the `ONLY_ACTIVE_ARCH=YES` for ios, that would be better that the `--active-arch-only` flag is more universal.
- not sure whether expo-cli will wait emulators booting up and start building. if not, it would be good to know android cpu arch before we can call adb shell. i found a way to know emulators's cpu type by reading `$HOME/.android/avd/<NAME>/config.ini` and there's the `abi.type=arm64-v8a`.

**Comment by alanjhughes** [comment]
Thanks for your feedback @Kudo! 
- This is a dev build only feature so I don't think we need to do anything for iOS as this is already the default there
Screenshot 2024-01-31 at 09 32 53

- The CLI will wait for the device to boot so if we resolve the device before the gradle props, we should always have access to it then when we check the architectures.

File: packages/@expo/cli/src/run/android/resolveGradlePropsAsync.ts [hunk 2]
```diff
@@ -34,11 +42,35 @@ export function resolveGradleProps(
...
+    return abis.join(',');
+  }
+
+  const devices = await getAttachedDevicesAsync();
+  const abisPromises = devices.map(async (device) => await getDeviceABIsAsync(device));
+  const abis = (await Promise.all(abisPromises))
+    .flat()
+    .filter((abi, i, arr) => arr.indexOf(abi) === i);
+  return abis.join(',');
+}
```
Comment (byCedric) [inline packages/@expo/cli/src/run/android/resolveGradlePropsAsync.ts:76]: I would make the `device` param required, `resolveDeviceAsync` should [always return an `AndroidDeviceManager` instance](https://github.com/expo/expo/blob/main/packages/%40expo/cli/src/run/android/resolveDevice.ts#L7-L23). And with that, we can simplify to:

```suggestion
async function getConnectedDeviceABIS(
  device: Device,
  buildType: string,
  allArch?: boolean
): Promise<string> {
  // Follow the same behavior as iOS, only enable this for debug builds
  if (allArch || buildType !== 'debug') {
    return '';
  }

  const abis = await getDeviceABIsAsync(device);
  return abis.join(',');
}
```

File: packages/@expo/cli/src/run/android/index.ts [hunk 3]
```diff
@@ -16,6 +16,7 @@ export const expoRunAndroid: Command = async (argv) => {
     '--no-install': Boolean,
     '--no-bundler': Boolean,
     '--variant': String,
+    '--all-arch': Boolean,
```
Comment (byCedric) [inline packages/@expo/cli/src/run/android/index.ts:19]: Let's add a clarification here, or in the help output generation.

```suggestion
    // Unstable, temporary fallback to disable active archs only behavior
    // TODO: replace with better fallback option, like free-form passing gradle props
    '--all-arch': Boolean,
```

**Review by byCedric — APPROVED** [review]
Awesome, thanks! Left 2 comments, but nothing major or merge-blocking.

**Pull request merged**